### PR TITLE
kvstoremesh: correctly remove cached data upon cluster disconnection

### PR DESCRIPTION
**Pull request by giorio94**
Currently, a comment attached to the Remove method claims that the cached data would have disappeared due to lease expiration. However, the lease is never expected to expire, as we use a single local client for all remote clusters. Additionally, depending on lease expiration would leave a significantly long race condition window in which stale data would still be present if the same cluster were to be reconnected.

In most cases, stale cached data is not a problem, as the etcd instance is stateless. However, it can create problems if the same cluster is first disconnected and then subsequently reconnected, without restarting the clustermesh-apiserver pod in between. Indeed, in this case we would have leftover entries that would not be cleaned up before reconnecting.

Let's address this issue by explicitly deleting the cached data upon cluster disconnection. This operation is performed synchronously (i.e., blocking other connection/disconnection operations) to prevent race conditions causing the deletion of incorrect data. Yet, to avoid blocking forever in case of errors, we only allow a limited number of retries, and give up with a loud error if they all fail.

Fixes: #issue-number

```release-note
Correctly remove data cached by KVStoreMesh for a given cluster when disconnecting from such cluster
```

### CODE REVIEW

File: pkg/clustermesh/kvstoremesh/remote_cluster.go
Comment (marseel) [inline pkg/clustermesh/kvstoremesh/remote_cluster.go]: nit: Is there any reason why we are overriding prefixes/keys?
I know it's technically correct, but a bit strange to read.

File: pkg/clustermesh/kvstoremesh/remote_cluster.go
Comment (marseel) [inline pkg/clustermesh/kvstoremesh/remote_cluster.go:188]: If I remember correctly, order in which we remove here is probably very important. 
I would love to see some reasoning of what happens to running cilium-agents and bootstrapping cilium-agents. 

For example, if we remove `ClusterConfigPrefix` first, then new clustermesh clients won't know that they should use syncCanaries, which could result in reading a partial state - which probably is fine, but it makes me wonder also about other cases.

**Review by marseel — CHANGES_REQUESTED** [review]
one nit, one bigger comment. 
Anyway, if we figure out these corner cases, I would encourage to do it behind a hidden flag, just in case, we need to disable it due to some race-condition.

File: pkg/clustermesh/kvstoremesh/remote_cluster.go
Comment (giorio94) [reply pkg/clustermesh/kvstoremesh/remote_cluster.go]: The idea is to avoid attempting to delete again successful prefixes/keys in case an error has occurred at a later stage. That's only a minor optimization, I can definitely drop it if you prefer.

File: pkg/clustermesh/kvstoremesh/remote_cluster.go
Comment (giorio94) [reply pkg/clustermesh/kvstoremesh/remote_cluster.go:188]: IMO deleting the cluster config first is the right thing to do here, because that effectively prevents bootstrapping agents from connecting (because they expect the config to be present), which is the expected behavior considering that we are deleting the cached data for that cluster in any case.

The deletion order of the other items doesn't matter once the agents are already connected, and they'll receive the deletion events for the various keys. Regardless, it is expected that agents will disconnect from the same cluster as well, hence they would drain all the keys in any case.

Another possibility would be that the bootstrapping agent already retrieved the cluster config, but hasn't started watching yet. In this case, deleting the synced prefix first prevents again from connecting, which looks correct to me, although that's arguably a rare condition.

**Comment by giorio94** [comment]
> one nit, one bigger comment. Anyway, if we figure out these corner cases, I would encourage to do it behind a hidden flag, just in case, we need to disable it due to some race-condition.

Thanks! I've added the hidden flag, PTAL.

**Review by marseel — APPROVED** [review]
Thanks, looks great!

**Comment by giorio94** [comment]
Converting back to draft, because this change introduces a possible race condition when disabling KVStoreMesh, and causes flakiness in the Cilium Cluster Mesh upgrade workflow. 

The reason being that disabling kvstoremesh via helm:
1. Switches the Cilium Agents clustermesh configuration from the local cache to the remote etcd;
2. Deletes the kvstoremesh container, causing the restart of the clustermesh-apiserver pod;
3. Deletes the kvstoremesh clustermesh configuration.

However, if the old kvstoremesh container is still running when the kvstoremesh configuration removal is propagated by the kubelet, it will interpret it as a cluster disconnection, hence triggering the newly introduced draining of the cached data. However, if the Cilium Agents are still watching this replica at that time (i.e., they have not yet observed the corresponding clustermesh configuration change), they will observe deletion events for keys that are actually still valid, hence causing the disruption of existing connections.

**Comment by giorio94** [comment]
The above issue is addressed waiting for a grace period (hardcoded to 3 minutes) before triggering the deletion of the cached data. This time window is meant to allow all Cilium agents to disconnect from KVStoreMesh in the meanwhile, both to reduce the overhead on etcd (as the deletions would not be watched by anyone) and prevent the race condition when disabling KVStoreMesh altogether. To address the fact that the removal process now takes longer, the first commit makes this operation asynchronous, so that additions and removals of other cluster configurations can be processed in parallel. Differently, the recreation of a configuration for a cluster which is being removed is queued and replayed once the removal completes, to still ensure the correct ordering of the operations.

File: pkg/clustermesh/common/clustermesh.go
Comment (marseel) [inline pkg/clustermesh/common/clustermesh.go:137]: As you mention it below:
> // Run onRemove in a separate go routing as potentially slow, to avoid
> // blocking the processing of further events in the meanwhile.

We might need to wait quite some time on `cm.wg.Wait()`
I'm not sure this is a good idea, agent shutdown should probably not block on clustermesh's deleted cluster cleanup. 
I know it was like that before too, but probably we shouldn't wait on it during shutting down of agent at all?

Currently, it seems like we setting [terminationGracePeriodSeconds to 1](https://github.com/cilium/cilium/blob/c22eceb75bb4187b011bc665058adc4aba765b78/install/kubernetes/cilium/values.yaml#L142) :exploding_head: 
So while I think for clustermesh it doesn't matter that much if it will be cleaned or not, I wonder if for other components it might actually matter whether they execute or not as they execute sequentially:
https://github.com/cilium/cilium/blob/2311f3d7811c8950b7f5e1e2ac290f8216be0dc3/vendor/github.com/cilium/hive/cell/lifecycle.go#L137

File: pkg/clustermesh/common/clustermesh.go
Comment (marseel) [inline pkg/clustermesh/common/clustermesh.go]:
```suggestion
		// The configuration for this cluster has been recreated before the cleanup
```

File: pkg/clustermesh/common/clustermesh.go
Comment (marseel) [inline pkg/clustermesh/common/clustermesh.go:241]: I'm not sure if doing this async doesn't make it worse for the case when someone changes cluster-name, without changing ClusterID, no?
Previous cluster is going to be deleting while config with a new cluster name will fail on ClusterID reservation?

**Review by marseel — CHANGES_REQUESTED** [review]
Overal makes sense, but I would question whether this part:
```
 To address the fact that the removal process now takes longer, the first commit makes this operation asynchronous, so that additions and removals of other cluster configurations can be processed in parallel. Differently, the recreation of a configuration for a cluster which is being removed is queued and replayed once the removal completes, to still ensure the correct ordering of the operations.
```
is necessary. Isn't that also increase the load on agent as we will be doing removal in parallel now? 

I'm honestly on the edge what would be better/safer. I don't think we should have super fast handling for cluster removal even if it blocks adding new clusters, if it introduces significant mental complexity around tombstones and removal of clusters.

This was mostly triggered due to the fact that I'm not really sure if using name vs clusterID in tombstones is a good idea.

**Comment by giorio94** [comment]
> Overal makes sense, but I would question whether this part:
> 
> ```
>  To address the fact that the removal process now takes longer, the first commit makes this operation asynchronous, so that additions and removals of other cluster configurations can be processed in parallel. Differently, the recreation of a configuration for a cluster which is being removed is queued and replayed once the removal completes, to still ensure the correct ordering of the operations.
> ```
> 
> is necessary. Isn't that also increase the load on agent as we will be doing removal in parallel now?
> 
> I'm honestly on the edge what would be better/safer. I don't think we should have super fast handling for cluster removal even if it blocks adding new clusters, if it introduces significant mental complexity around tombstones and removal of clusters.
> 

Generally speaking I was also a bit on the fence about the best approach here (and I'm not a big fan of the new logic, which is less straightforward). Yet, I think that there are a couple of main aspects justifying the asynchronous deletion, although a bit more complex.

Before that, as a quick recap, the logic to observe configuration changes is used both by Cilium agents and kvstoremesh (and also by the operator when endpointslice synchronization is enabled, but that's akin to the agents). In both cases, upon new cluster detection (i.e., appearance of a new configuration), we start a controller that runs in background, establishes the connection to the target etcd instance, retrieves the cluster configuration (i.e., the one containing the ClusterID) and eventually calls the agent/kvstoremesh specific logic to mainly start the appropriate watchers (the agents perform the ClusterID validation here). Upon configuration removal, instead the logic stops the above controller (i.e., stopping the watchers and closing the etcd client) and then calls the specific removal logic. Agents essentially emit a synthetic deletion event for all known entries, to drain them. Differently, with this PR, kvstoremesh triggers the cleanup of the prefixes (with the initial grace period to avoid the issues discussed previously). All the disconnect and removal part was historically performed synchronously, and is changed by the first commit.

Back to the reasons,
* Making the deletion asynchronous brings it on par with the connection to a new cluster, which is also performed asynchronously via a controller; 
* Blocking the configuration of a new cluster for a relatively long period of time would be confusing for users. Let's consider the kvstoremesh case with the 3 minutes grace period. Supposing a user disconnects 5 clusters and connects a new one, the new one may not be reconciled for 15 minutes, giving the impression that something didn't work. The order in which the cluster configuration events are processed is also undeterministic (it depends on the inotify), which makes it even worse.
* Most importantly, it would lead to asymmetry between the agents and kvstoremesh. Back to the previous example, and besides the fact that ordering may be different, the removal on the agent side would be relatively fast, while the grace period would slow down the kvstoremesh cleanup quite a lot. That would mean that agents would attempt connecting to kvstoremesh for the new cluster way earlier than when kvstoremesh could start performing the synchronization, hence entering into a backoff loop as the cluster configuration has not yet been propagated.

> Isn't that also increase the load on agent as we will be doing removal in parallel now?

Agents do need to process the synthetic deletion events upon cluster removal, so this can potentially increase the load a bit. However, that's not dissimilar to the initial listing upon agent startup, which is also performed in parallel, and it is arguably more demanding (both in terms of operations, and because it involves all the clusters the agents are connected to). As for kvstoremesh, instead, the expectation is that all agents have already stopped the watchers at that point, so the deletion of the prefixes should be relatively cheap.

> This was mostly triggered due to the fact that I'm not really sure if using name vs clusterID in tombstones is a good idea.

IMO cluster names are correct there, because they correspond to the names of the configurations. Hence, that ensures that subsequent creations and removals of the same configuration file are handled correctly, and coalescenced as appropriate. Differently, ClusterID are only retrieved after connecting to the etcd instance.

File: pkg/clustermesh/common/clustermesh.go
Comment (giorio94) [reply pkg/clustermesh/common/clustermesh.go:137]: I think we need to distinguish different cases here:

* Plain agent or kvstoremesh shutdown, with no clustermesh configurations removals (i.e., the standard case): clustermesh just stops the watchers and closes the etcd client. This was, and continues to be, performed synchronously by the stop hook, and is expected to be instantaneous (unless the agent is blocked processing events already received).
* agents shutdown while the removal of one or more clusters is being processed. In this case, it is possible that the cleanup doesn't complete in time with such a short termination grace period, as the agents would need to finish processing all the synthetic deletion events. However, this was already the case earlier, and it is an extremely rare condition IMO.
* kvstoremesh shutdown while the removal of one or more clusters is being processed. In that case, [the context of the job should be closed](https://github.com/giorio94/cilium/blob/mio%2Fkvstoremesh-delete-cluster/pkg/clustermesh/kvstoremesh/remote_cluster.go#L183), hence causing the removal logic to return immediately, and unblock the wait group. Now that I'm writing this, I reckon there may be a startup/shutdown ordering issue that makes that not work as expected, let me double check and address it in case.

File: pkg/clustermesh/common/clustermesh.go
Comment (giorio94) [reply pkg/clustermesh/common/clustermesh.go:241]: I don't think there would be any difference, as the inotify watcher is non deterministic (as in, it may be, but it depends on the actual file names), and we could either first observe the addition or deletion event.

Comment (marseel) [reply pkg/clustermesh/common/clustermesh.go:241]: makes sense, thanks for the explanation.

**Review by marseel — APPROVED** [review]
I want to make sure we understand all the consequences of this change first - especially as this change touches both kvstoremesh and the agent side, before merging :) 

I would say that two examples that you gave:
```
agents shutdown while the removal of one or more clusters is being processed. In this case, it is possible that the cleanup doesn't complete in time with such a short termination grace period, as the agents would need to finish processing all the synthetic deletion events. However, this was already the case earlier, and it is an extremely rare condition IMO.
```
and
```
Blocking the configuration of a new cluster for a relatively long period of time would be confusing for users. Let's consider the kvstoremesh case with the 3 minutes grace period. Supposing a user disconnects 5 clusters and connects a new one, the new one may not be reconciled for 15 minutes, giving the impression that something didn't work. The order in which the cluster configuration events are processed is also undeterministic (it depends on the inotify), which makes it even worse.
```
have probably a similar frequency of happening. In general, adding/removing clusters does not happen often, but once it happens, churn on nodes is not something very unlikely to happen during this time, nodes come and go all the time.

I know it was like this before - blocking agent on cluster removal, so it's not something blocking ( :wink: )  this PR, but perhaps something to improve later on.

**Comment by giorio94** [comment]
> I would say that two examples that you gave:
> 
> ```
> agents shutdown while the removal of one or more clusters is being processed. In this case, it is possible that the cleanup doesn't complete in time with such a short termination grace period, as the agents would need to finish processing all the synthetic deletion events. However, this was already the case earlier, and it is an extremely rare condition IMO.
> ```
> 
> and
> 
> ```
> Blocking the configuration of a new cluster for a relatively long period of time would be confusing for users. Let's consider the kvstoremesh case with the 3 minutes grace period. Supposing a user disconnects 5 clusters and connects a new one, the new one may not be reconciled for 15 minutes, giving the impression that something didn't work. The order in which the cluster configuration events are processed is also undeterministic (it depends on the inotify), which makes it even worse.
> ```
> 
> have probably a similar frequency of happening. In general, adding/removing clusters does not happen often, but once it happens, churn on nodes is not something very unlikely to happen during this time, nodes come and go all the time.

Yep, I agree with that. Currently the effect of possibly blocking the shutdown in that situation is mostly negligible due to the termination grace period being set to 1 second, but I agree that it could be worth an improvements if we decide to raise that period. One possibility could be using a context to interrupt the generation of the synthetic deletion events on shutdown (the GC on restart is expected to take care of them regardless).

File: pkg/clustermesh/common/clustermesh.go
Comment (giorio94) [reply pkg/clustermesh/common/clustermesh.go:137]: > I reckon there may be a startup/shutdown ordering issue that makes that not work as expected, let me double check and address it in case.

Fixed with the latest push: https://github.com/cilium/cilium/compare/81721f9e913c55d75137a74c4bb0ec45d5815fc2..a5ad5df50a339422506088edab0b67c68336715a

**Review by marseel — COMMENTED** [review]
Nice, looks good. 

A bit confusing for me was that onRemove uses a different context from drainJobFn though and we don't cancel drainJobFn in Stop() of `func (cm *clusterMesh) Stop(cell.HookContext) error {`
If my understanding is correct these context are attached to two different Stop hooks, right? One from jobGroup thing, the other one to new rctx? I know it's still correct, but a bit confusing to me. 
Ideally, we would have rctx attached to jobGroup, no?

Just wanted to confirm my understanding, but lgtm otherwise.

**Comment by giorio94** [comment]
> Nice, looks good.
> 
> A bit confusing for me was that onRemove uses a different context from drainJobFn though and we don't cancel drainJobFn in Stop() of `func (cm *clusterMesh) Stop(cell.HookContext) error {` If my understanding is correct these context are attached to two different Stop hooks, right? One from jobGroup thing, the other one to new rctx? I know it's still correct, but a bit confusing to me. Ideally, we would have rctx attached to jobGroup, no?
> 
> Just wanted to confirm my understanding, but lgtm otherwise.

Yes, the reason for the separate `rctx` is that the job gets registered earlier, and hence its stop hook is called later than `clustermesh.Stop`. Specifically, the job is automatically provided and registered by hive given that we retrieve it as a parameter [here](https://github.com/giorio94/cilium/blob/mio%2Fkvstoremesh-delete-cluster/pkg/clustermesh/kvstoremesh/kvstoremesh.go#L79), and the resulting ordering is:

```
Start hooks:

  ...
  • *job.group.Start (kvstoremesh.kvstoremesh)
  • *kvstoremesh.KVStoreMesh.Start (kvstoremesh.kvstoremesh)
  • *common.clusterMesh.Start (kvstoremesh.kvstoremesh)
  ...

Stop hooks:

  ...
  • *common.clusterMesh.Stop (kvstoremesh.kvstoremesh)
  • *kvstoremesh.KVStoreMesh.Stop (kvstoremesh.kvstoremesh)
  • *job.group.Stop (kvstoremesh.kvstoremesh)
  ...
```

Another possible approach would be manually constructing the job group from a job registry, and registering it below the other ones, [here](https://github.com/giorio94/cilium/blob/mio%2Fkvstoremesh-delete-cluster/pkg/clustermesh/kvstoremesh/kvstoremesh.go#L99-L103). That's how I addressed it initially. However, it seemed to me that having an explicit "shutting-down" signal via the `Remove` context may be helpful as well to potentially interrupt the draining process in the agents, if needed. 

I can revert to the other approach if you prefer that though. Another possibility could be propagating `rctx` into `drainJobFn`, and use that rather than the job context, although it doesn't seem great either to me. WDYT?

**Comment by giorio94** [comment]
After an offline discussion with Marcel, I ended up reworking the last commit to remove the usage of the job, which turned out being tricky to reason about, in favor of bespoke logic. I've also reworked the tests to use a fake clock, to actually test the different timings for the grace period and backoff.

**Comment by giorio94** [comment]
Rebased onto main to hopefully fix the ci-ces-migrate failure.

**Review by marseel — APPROVED** [review]
very nice and easy to read now, double approve :100:

**Comment by giorio94** [comment]
Marked for backport to v1.16, given that kvstoremesh is no longer beta, and this fix missed the branch window.

File: pkg/clustermesh/kvstoremesh/kvstoremesh.go
Comment (joestringer) [inline pkg/clustermesh/kvstoremesh/kvstoremesh.go:49]: Please prefer `enable-drain-on-disconnection` rather than `disable-drain-on-connection`. Defining options as negative is confusing in some languages due to the double negative (if you _don't disable_ foo, ie `--disable-foo=false`, does that mean it's enabled or just that it's not explicitly enabled?)

Comment (joestringer) [reply pkg/clustermesh/kvstoremesh/kvstoremesh.go:49]: Alternative that may even be better: don't use `enable/disable` at all.

Comment (marseel) [reply pkg/clustermesh/kvstoremesh/kvstoremesh.go:49]: I think we can still change the name of it - it is a hidden option as we did not expect anyone to use it, just as a fail-safe mechanism. 
But that's a really good point with naming, I was scratching my head a few times before with similar options too.

Comment (giorio94) [reply pkg/clustermesh/kvstoremesh/kvstoremesh.go:49]: > Please prefer enable-drain-on-disconnection rather than disable-drain-on-connection. Defining options as negative is confusing in some languages due to the double negative (if you don't disable foo, ie --disable-foo=false, does that mean it's enabled or just that it's not explicitly enabled?)

Hmm, I see your point. 

The reason for calling it `disable` is to avoid `true` by default boolean options, which also tend to be confusing and problematic in certain contexts (ie., a user may expect that entirely omitting `--enable-drain-on-disconnection=false` would be equivalent to specifying it, which would not be the case defaulting to true -- conversely --disable-drain-on-disconnection looks more intuitive to me). Although a bit unrelated, AFAIR a few true by default settings are also currently broken in the helm chart for similar reasons.

That said, as Marcel mentioned, this option is a temporary fail-safe (hidden) toggle not intended to be exposed to users, and I agree that we should still be able to change the name without particular consequences in case.

Comment (joestringer) [reply pkg/clustermesh/kvstoremesh/kvstoremesh.go:49]: >  [...] a user may expect that entirely omitting --enable-drain-on-disconnection=false would be equivalent to specifying it, which would not be the case defaulting to true [...]

To a degree I get the concern around default-true boolean options, but that's also just inherently how features work - features are either enabled or disabled. Depending on the maturity of the feature and how useful it is to the average user, we may decide to enable or disable different flags by default. There's then two important aspects to the UX - can users easily understand the default posture of the setting? And can they override it themselves?

Here's where I'm coming from in context of this specific flag: If I specify `--disable-drain-on-disconnection=false`, that means: Please don't "Do not drain cached data upon cluster disconnection".

This point is not a big deal for hidden options, but I raise it to encourage discussion about how we can provide more consistent UX across different flags. My gut feeling is that unless there is a really strong argument in a specific scenario, using "disable" (or other negative modifiers) in a flag name is almost always more confusing than the reverse.

**Pull request merged**
